[AArch64][SME] Remove immediate argument restriction for svldr and svstr

SamTebbs33 opened the pull request:
The svldr_vnum and svstr_vnum builtins always modify the base register and tile slice and provide immediate offsets of zero, even when the offset provided to the builtin is an immediate. This patch optimises the output of the builtins when the offset is an immediate, to pass it directly to the instruction and to not need the base register and tile slice updates.

SamTebbs33 committed:
[AArch64][SME] Remove immediate argument restriction for svldr and svstr

The svldr_vnum_za and svstr_vnum_za builtins/intrinsics currently
require that the vnum argument be an immediate, since the instructions
take an immediate vector number. However, we emit 0 as the immediate
for the instruction no matter what, and instead modify the base register.

This patch removes that restriction on the argument, so that the
argument can be a non-immediate. If an appropriate immediate was
passed to the builtin then CGBuiltin passes that directly to the LLVM
intrinsic, otherwise it modifies the base register as is existing
behaviour.

SamTebbs33 committed:
fixup: remove erroneously included file

kmclaughlin-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -9665,18 +9665,34 @@ Value *CodeGenFunction::EmitSMEZero(const SVETypeFlags &TypeFlags,
+      llvm::Value *CntsbCall = Builder.CreateCall(Cntsb, {}, "svlb");
+
+      llvm::Value *VecNum = Ops[2];
+      llvm::Value *MulVL = Builder.CreateMul(
+          CntsbCall,
+          VecNum,
+          "mulvl");
```
This should fit into a single line, please can you run clang-format on the patch?

SamTebbs33 replied:
I did forget to do that. Will do.

kmclaughlin-arm commented on clang/test/CodeGen/aarch64-sme-intrinsics/acle_sme_ldr.c:
```diff
@@ -8,7 +8,7 @@
 // CHECK-C-LABEL: @test_svldr_vnum_za(
 // CHECK-CXX-LABEL: @_Z18test_svldr_vnum_zajPKv(
 // CHECK-NEXT:  entry:
-// CHECK-NEXT:    tail call void @llvm.aarch64.sme.ldr(i32 [[SLICE_BASE:%.*]], ptr [[PTR:%.*]])
```
Were these changes to SLICE_BASE and PTR intentional? When I try to run this test now it fails with the following:
`undefined variable: SLICE_BASE` &
`undefined variable: PTR`

SamTebbs33 replied:
These changes actually came from the test update script llvm/utils/update_cc_test_checks.py, but I can revert them as they weren't intentional changes.

kmclaughlin-arm commented on clang/test/CodeGen/aarch64-sme-intrinsics/acle_sme_str.c:
```diff
@@ -8,31 +8,28 @@
 
-// CHECK-C-LABEL: @test_svstr_vnum_za_1(
-// CHECK-CXX-LABEL: @_Z20test_svstr_vnum_za_1jPv(
+// CHECK-C-LABEL: define dso_local void @test_svstr_vnum_za_1(
```
Since the other tests don't check for `define dso_local void`, I don't think it's necessary here

SamTebbs33 replied:
Absolutely, thanks for spotting these.

kmclaughlin-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -9665,18 +9665,34 @@ Value *CodeGenFunction::EmitSMEZero(const SVETypeFlags &TypeFlags,
-  }
+  if (Ops.size() == 2) {
+    // Intrinsics without a vecnum also use this function, so just provide 0
+    Ops.push_back(Ops[1]);
+    Ops[1] = Builder.getInt32(0);
```
Just a suggestion, but the extra operand could be added with insert, e.g.
`Ops.insert(Ops.begin() + 1, Builder.getInt32(0));`

SamTebbs33 replied:
That's much better, thanks.

kmclaughlin-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -9665,18 +9665,34 @@ Value *CodeGenFunction::EmitSMEZero(const SVETypeFlags &TypeFlags,
+    // Intrinsics without a vecnum also use this function, so just provide 0
+    Ops.push_back(Ops[1]);
+    Ops[1] = Builder.getInt32(0);
+  } else {
+    int Imm = -1;
+    if (ConstantInt* C = dyn_cast<ConstantInt>(Ops[2]))
+      if (C->getZExtValue() <= 15)
+          Imm = C->getZExtValue();
+
+    if (Imm != -1) {
+      Ops[2] = Ops[1];
+      Ops[1] = Builder.getInt32(Imm);
```
Would it be possible to remove the check for -1 with something like this?
```
} else if (ConstantInt* C = dyn_cast<ConstantInt>(Ops[2])) {
  auto Imm = C->getZExtValue();
  if (Imm <= 15) {
    Ops[2] = Ops[1];
    Ops[1] = Builder.getInt32(Imm);
}
```

SamTebbs33 replied:
The thing is that I want the rest of the code to run if the argument is an immediate but it isn't in the range of 0 - 15. I could move the <= 15 check to where `Imm != -1` is, I think that would be cleaner.

efriedma-quic commented:
This patch seems sort of contradictory to me.  You're encouraging users to pass a non-constant slice_offset, presumably because we expect the backend to optimize it... but at the same time, you're special-casing constant slice_offsets because you don't expect the backend to optimize it.  What do you actually expect the backend to do in practice?

I don't see any backend tests involving a non-zero slice_offset?

We currently have some code in getMemVTFromNode() in AArch64ISelDAGToDAG.cpp that handles sme_ldr/sme_str.  Is that code currently reachable?  If it is reachable, does it need to be adjusted for this change?

sdesmalen-arm reviewed (commented):
The changes to the LLVM IR intrinsics are to better utilise the reg+imm form of the instructions, but they aren't required to remove the immediate-argument restriction for svldr/svstr. I think it would be better to split the PR into two PRs:
* One patch that fixes `vnum` (uint64_t -> int64_t) and removes the 'must be immediate argument' restriction.
* Another patch that tries to better use the immediate in the instruction by changing the LLVM IR intrinsic with an extra parameter.

SamTebbs33 commented on clang/include/clang/Basic/arm_sme.td:
```diff
@@ -44,10 +44,9 @@ defm SVLD1_ZA32 : ZALoad<"za32", "i", "aarch64_sme_ld1w", [ImmCheck<0, ImmCheck0
 defm SVLD1_ZA128 : ZALoad<"za128", "q", "aarch64_sme_ld1q", [ImmCheck<0, ImmCheck0_15>]>;
 
-def SVLDR_VNUM_ZA : MInst<"svldr_vnum_za", "vmQi", "",
+def SVLDR_VNUM_ZA : MInst<"svldr_vnum_za", "vmQn", "",
```
Right. It currently takes an unsigned integer but I can change it to be signed as per your suggestion above.

dtemirbulatov commented on llvm/test/CodeGen/AArch64/sme-intrinsics-loads.ll:
```diff
@@ -292,7 +310,7 @@ define void @ldr_with_off_16mulvl(ptr %ptr) {
   %mulvl = mul i64 %vscale, 256
   %base = getelementptr i8, ptr %ptr, i64 %mulvl
-  call void @llvm.aarch64.sme.ldr(i32 16, ptr %base)
+  call void @llvm.aarch64.sme.ldr(i32 16, i32 0, ptr %base)
```
Maybe test non zero second parameter to see how it apllied in LDR instruction?

SamTebbs33 commented:
I've decided to keep this PR up for the patch that modifies the codegen and IR intrinsic since there's a lot of good relevant review of that here. I've made a [new PR](https://github.com/llvm/llvm-project/pull/68908) with the separated changes that just modify the clang builtin. This PR will later be rebased on top of that PR after it's merged.

SamTebbs33 committed:
fixup! Use DAGToDAG approach

SamTebbs33 commented:
I've updated this patch with my progress in lowering the intrinsics in DAGToDAG, hopefully capturing your feedback in the process.

efriedma-quic commented on llvm/test/CodeGen/AArch64/sme-intrinsics-loads.ll:
```diff
@@ -292,7 +292,42 @@ define void @ldr_with_off_16mulvl(ptr %ptr) {
+; CHECK-NEXT:    rdsvl x8, #1
+; CHECK-NEXT:    mov w12, #16 // =0x10
+; CHECK-NEXT:    madd x8, x8, x12, x0
+; CHECK-NEXT:    ldr za[w12, 0], [x8]
```
This looks suspicious; shouldn't w12 contain 16+16=32?

SamTebbs33 replied:
I think you're right. It looks like the same thing seems to be happening with `ldr_with_off_var` as well.

efriedma-quic reviewed (commented):
Could you briefly comment on the tradeoff of adding an argument to the intrinsic, vs. pattern-matching constant offsets to the existing intrinsic?

efriedma-quic commented:
Also, you might want to consider marking the offset immarg, instead of trying to handle variable offsets in isel.

sdesmalen-arm commented:
> Also, you might want to consider marking the offset immarg, instead of trying to handle variable offsets in isel.

`vnum` doesn't need to be an immediate in the C/C++ intrinsic, so it makes sense to just pass it onto LLVM so that ISel can try to fold it into the immediate operands of the instruction. The reason for having a separate operand for this in the intrinsic, is because the tile-slice offset and the (scaled) pointer offset *must* be the same, and if we encode that as a separate operand ISel can simply ensure that the same immediate is used. At the moment it's more difficult to match the immediate (because we need to analyse both the pointer and the tileslice), so have simply resorted to always use an immediate of `0`.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -1741,6 +1742,54 @@ void AArch64DAGToDAGISel::SelectCVTIntrinsic(SDNode *N, unsigned NumVecs,
+  if (auto ImmNode = dyn_cast<ConstantSDNode>(VecNum))
+    Imm = ImmNode->getZExtValue();
+
+  if (Imm >= 0 && Imm <= 15) {
```
> Something still needs to happen with the immediate argument when it doesn't fit entirely into the instruction's immediate operand. The current result is not correct.

I don't quite understand what that something is if immediate if it doesn't fit in the range. Currently the immediate is multiplied by svl and added to the base register if it doesn't fit. What else should be done if it doesn't fit in the range?

> Maybe you can restructure the code in such a way that it first tries to fold as much of the constant into the instruction's immediate operand, and then adds the remainder to the tile-slice and to the pointer (multiplied by SVL), as you do in the code below.

Do you mean that if, for example, we have an immediate vecnum of 17, then we put 15 into the offset field of the instruction and add 2 x SVL to the base and slice registers? Naively I don't see that being an improvement as you're doing the multiplication in two places (ldr instruction and an extra madd) rather than one.

SamTebbs33 replied:
We had an offline discussion and agreed that when the vnum is an immediate but is greater than a multiple of 15, it makes sense to modify the base and slice registers by that multiple and put the remainder in the immediate field of the instruction. That way when we unroll a loop with lots of ldrs offset by one from each other, they can all share the base and slice modifications but have increasing immediates.

SamTebbs33 commented:
I've changed the approach to consider immediates outside of 0-15 and fixed the issue of the tile slice not being updated. Please let me know what you think.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -1741,6 +1742,69 @@ void AArch64DAGToDAGISel::SelectCVTIntrinsic(SDNode *N, unsigned NumVecs,
+      Remainder = CurDAG->getTargetConstant(Imm, DL, MVT::i32);
+      NeedsAdd = false;
+    } else {
+      Remainder = CurDAG->getTargetConstant(Imm % 15, DL, MVT::i32);
```
"% 15"?  Not 16?

SamTebbs33 replied:
Yep, the range of the immediate is 0-15.

efriedma-quic replied:
`Imm % 16` is in the range 0-15.  `Imm % 15` is in the range 0-14.

SamTebbs33 replied:
You are right, thank you. I've updated it to modulo 16 instead and added some tests for higher immediates.

efriedma-quic reviewed (commented):
The resulting code here still isn't great quality for the case where the constant isn't folded.  Maybe worth considering doing the lowering earlier (DAGCombine?  Maybe even a late IR optimization?), so the lowered arithmetic can be optimized.

SamTebbs33 committed:
fixup! lower in ISelLowering instead

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -4825,6 +4827,72 @@ SDValue AArch64TargetLowering::getPStateSM(SelectionDAG &DAG, SDValue Chain,
+    // operand is an immediate in range
+    if (auto ImmNode = dyn_cast<ConstantSDNode>(VecNum.getOperand(1))) {
+      int Imm = ImmNode->getSExtValue();
+      if (Imm >= 0 && Imm <= 15) {
```
Can you write things in such a way that the immediate is always handled the same way, regardless of whether it came from the RHS of an ADD, or whether it comes directly from the `N->getOperand(4)`? For example, for the case where `VecNum = add %vnum, 17`, it could still use an immediate of `1` and add the `rdsvl * (vnum + 16)` to the baseptr and `(vnum + 16)` to the tileslice value.

SamTebbs33 replied:
My latest two commits do this, which improves the codegen as well since we properly fold successive loads/stores with adds of large immediates. See the test [here](https://media.tenor.com/vYhPH4gA184AAAAC/vroom-car.gif). Thanks for the idea!

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -4825,6 +4827,72 @@ SDValue AArch64TargetLowering::getPStateSM(SelectionDAG &DAG, SDValue Chain,
+    TileSlice = DAG.getNode(ISD::ADD, DL, MVT::i32, SliceAddOps);
+  }
+
+  SmallVector<SDValue, 4> Ops = {N.getOperand(0), TileSlice, Base, Remainder};
```
```suggestion
  SmallVector<SDValue, 4> Ops = {/*Chain=*/N.getOperand(0), TileSlice, Base, Remainder};
```

SamTebbs33 committed:
fixup! Update check lines

SamTebbs33 commented on llvm/test/CodeGen/AArch64/sme-intrinsics-loads.ll:
```diff
@@ -292,23 +292,101 @@ define void @ldr_with_off_16mulvl(ptr %ptr) {
+; CHECK-LABEL: ldr_with_off_var:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    // kill: def $w2 killed $w2 def $x2
+; CHECK-NEXT:    sxtw x8, w2
```
I did forget to stage those changes. Thanks for noticing.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -4825,6 +4827,72 @@ SDValue AArch64TargetLowering::getPStateSM(SelectionDAG &DAG, SDValue Chain,
+    // Just add vnum to the tileslice
+    SDValue BaseMulOps[] = {
+        SVL, VecNum.getValueType() == MVT::i32
+                 ? DAG.getNode(ISD::SIGN_EXTEND, DL, MVT::i64, VecNum)
```
We do indeed, I'll remove the condition.

SamTebbs33 committed:
fixup! Clean up node creation

SamTebbs33 committed:
fixup! modulo 16 instead of 15

SamTebbs33 committed:
fixup! move add check before range check

SamTebbs33 committed:
fixup! add some examples above the function

SamTebbs33 committed:
fixup! fix mlir test

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -4825,6 +4827,113 @@ SDValue AArch64TargetLowering::getPStateSM(SelectionDAG &DAG, SDValue Chain,
+  bool NeedsAdd = true;
+  auto ImmNode = dyn_cast<ConstantSDNode>(VecNum);
+  if (ImmNode || Addend != 0) {
+    int Imm = ImmNode ? ImmNode->getSExtValue() + Addend : Addend;
```
In this expression either (Addend is 0 and ImmNode is non-zero) or (ImmNode is zero and Addend is non-zero). I don't think there is an instance where they can both be non-zero.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -4825,6 +4827,113 @@ SDValue AArch64TargetLowering::getPStateSM(SelectionDAG &DAG, SDValue Chain,
+      // slice and base register. We could also get here because Addend != 0 but
+      // vecnum is not an immediate, in which case we still want the base and
+      // slice register to be modified
+      NeedsAdd = !ImmNode;
```
Maybe it's me, but I find this logic a little tricky to follow. Specifically here that the value for NeedsAdd depends on previous control flow, which depends on whether ImmNode is defined.

It might be a bit simpler to follow if you progressively break down VecNum in two subsequent steps.

First break it down into:
* A variable part (e.g. for `i + 17` that would be `i`)
* A constant (e.g. for `i + 17` that would be `17`)

Second to break down `17` into:
* A base constant (for `17` that would be `15`)
* An immediate (for `17` that would be `2`)

When you then fold the base constant into the variable part, you can avoid the need for `NeedsAdd` because it can be inferred from whether there is a variable part, e.g.:

```
// First split VecNum into a "Variable" and "Constant" part.
int32_t ConstAddend = 0;
SDValue VariableAddend = VecNum;
if (VecNum.getOpcode() == ISD::ADD && isa<ConstantSDNode>(VecNum.getOperand(1))) {
  ConstAddend = cast<ConstantSDNode>(VecNum.getOperand(1))->getSExtValue();
  VariableAddend = VecNum.getOperand(0);
} else if (auto ImmNode = dyn_cast<ConstantSDNode>(VecNum)) {
  ConstAddend = ImmNode->getSExtValue();
  VariableAddend = SDValue();
}

// Further try to split the constant into an immediate.
int32_t ImmAddend = ConstAddend % 16;
if (int32_t C = (ConstAddend - ImmAddend)) {
  SDValue CVal = DAG.getConstant(C, DL, MVT::i32);
  VariableAddend = VariableAddend ?
    DAG.getNode(ISD::ADD, DL, MVT::i32, {VariableAddend, CVal}) : CVal;
}

if (VariableAddend) {
  // Get the vector length that will be multiplied by VariableAddend
  auto SVL = DAG.getNode(AArch64ISD::RDSVL, DL, MVT::i64,
  ...
}
```

SamTebbs33 replied:
I actually didn't realise that `SDValue()` is a falsey value. That certainly does eliminate the need for the `NeedsAdd` boolean. Thank you!

SamTebbs33 committed:
simplify code

SamTebbs33 committed:
separate out SME_LDR_STR_Intrinsic as its now used by zt ldr/str

sdesmalen-arm reviewed (approved):
Thanks for all the changes! LGTM with few little nits addressed.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -4850,6 +4852,93 @@ SDValue AArch64TargetLowering::getPStateSM(SelectionDAG &DAG, SDValue Chain,
                      Mask);
 }
 
+// Lower an SME LDR/STR ZA intrinsic to LDR_ZA_PSEUDO or STR_ZA.
```
And it's no longer precise as we lower it to the SDNode first.

SamTebbs33 committed:
cleanup

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -4850,6 +4852,93 @@ SDValue AArch64TargetLowering::getPStateSM(SelectionDAG &DAG, SDValue Chain,
+  auto LdrStr =
+      DAG.getNode(IsLoad ? AArch64ISD::SME_ZA_LDR : AArch64ISD::SME_ZA_STR, DL,
+                  MVT::Other, Ops);
+  return LdrStr;
```
Thanks, I must have kept this from before the refactor.

SamTebbs33 commented:
> Thanks for all the changes! LGTM with few little nits addressed.

Thanks!

SamTebbs33 committed:
fix mlir test again
